pkg/egressgateway/manager.go

```diff
@@ -191,104 +152,93 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
-		return err
-	}
+func (manager *Manager) removeUnusedEgressRules() {
+	egressPolicies := map[egressmap.EgressPolicyKey4]egressmap.EgressPolicyVal4{}
+	egressmap.EgressPolicyMap.IterateWithCallback(
+		func(key *egressmap.EgressPolicyKey4, val *egressmap.EgressPolicyVal4) {
+			egressPolicies[*key] = *val
+		})
+
+nextPolicyKey:
+	for policyKey, policyVal := range egressPolicies {
+		for _, policyConfig := range manager.policyConfigs {
+			for _, endpoint := range manager.epDataStore {
+				if !policyConfig.selectsEndpoint(endpoint) {
+					continue
+				}
 
-	if endpointPolicies, ok := manager.policyEndpoints[epData.id]; ok {
-		for _, polID := range endpointPolicies {
-			if polID == config.id {
-				log.Debug("Endpoint to policy mapping already exists.")
-				return nil
+				for _, endpointIP := range endpoint.ips {
+					for _, dstCIDR := range policyConfig.dstCIDRs {
+						if policyKey.Match(endpointIP, dstCIDR) &&
+							policyVal.Match(policyConfig.egressIP, policyConfig.egressIP) {
+							continue nextPolicyKey
+						}
+					}
+				}
 			}
 		}
-		// Add policy to existing list
-		manager.policyEndpoints[epData.id] = append(manager.policyEndpoints[epData.id], config.id)
-	} else {
-		// Add policy to new list
-		pe := []policyID{config.id}
-		manager.policyEndpoints[epData.id] = pe
-	}
-	return nil
-}
 
-func (manager *Manager) updateEgressMap(ips []string, config *PolicyConfig) error {
-	for _, ip := range ips {
-		sip := net.ParseIP(ip).To4()
-		for _, dstCIDR := range config.dstCIDRs {
-			// As currently designed, the egressIP serves two purposes, one for forwarding traffic
-			// to the gateway node, the other for SNATing the egress traffic on the gateway.
-			err := egressmap.ApplyEgressPolicy(sip, *dstCIDR, config.egressIP, config.egressIP)
-			if err != nil {
-				return err
-			}
+		if err := egressmap.RemoveEgressPolicy(policyKey.GetSourceIP(), *policyKey.GetDestCIDR()); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.SourceIP:        policyKey.GetSourceIP(),
+				logfields.DestinationCIDR: policyKey.GetDestCIDR(),
+			}).Error("Error removing egress policy.")
 		}
 	}
-	return nil
 }
 
-func (manager *Manager) deleteEgressMap(config *PolicyConfig, epData *endpointMetadata) error {
-	for _, ip := range epData.ips {
-		sip := net.ParseIP(ip).To4()
-		for _, dstCIDR := range config.dstCIDRs {
-			err := egressmap.RemoveEgressPolicy(sip, *dstCIDR)
-			if err != nil {
-				return err
-			}
-		}
+// reconcile is responsible for reconciling the state of the manager (i.e. the
+// desired state) with the actual state of the node (egress policy map entries).
+//
+// Whenever it encounters an error, it will just log it and move to the next
+// item, in order to reconcile as many states as possible.
+func (manager *Manager) reconcile() {
+	if !manager.k8sCacheSyncedChecker.K8sCacheIsSynced() {
+		return
 	}
-	return nil
+
+	manager.addMissingEgressRules()
+	manager.removeUnusedEgressRules()
```

Comment:
The order of those two calls seems to matter for updated entries, as we would otherwise remove them before updating them. Is it worth a comment?

Reply:
Yep, order matters here, will add a comment :+1: